**Feature:** useSyncToPathState hook (#918)
contiamo/operational-ui: src/hooks/useURLState/README.md

It's quite convenient to be able to copy and paste a link to somebody to show them a particular page of your application.

It's even more convenient if this link has the state of the current page! This is what this hook is made for.

> Disclaimer: when we say synchronized, we don't mean two way binding. It will read initial state from URL on component mount, but after internal state will be source of truth. Which is ok most of the time, because we use `replaceState` and there is no way from user point of view to change URL without triggering page reload. The only edge case to be aware here is when you use `useURLState` with `pushState`, in this case you will need to make sure that component lives inside a valid router context.

Reviewer: ```suggestion
Disclaimer: when we say synchronized, we don't mean two way binding. It will read the initial state from the URL on component mount, but afterwards, this hook's internal state will be source of truth. 

This is ok most of the time because we use `replaceState` and there is no way from user's point of view to change the URL without triggering a page reload. 

The only edge case to be aware of here is when you use `useURLState` with `pushState`. In this case, you will need to make sure that this hook lives inside a valid router context.
```

> Disclaimer: when we say synchronized, we don't mean two way binding. It will read initial state from URL on component mount, but after internal state will be source of truth. Which is ok most of the time, because we use `replaceState` and there is no way from user point of view to change URL without triggering page reload. The only edge case to be aware here is when you use `useURLState` with `pushState`, in this case you will need to make sure that component will re-mount on push state, for example, if you use `react-router` use component with `useURLState` inside router. 

Reviewer: > in this case you will need to make sure that component will re-mount on push state, for example, if you use `react-router` use component with `useURLState` inside router.

This needs to be reworded. You need to make sure your component will remount?

Author: Only if you want two way binding, but it's not the purpose of this component 😉 The idea is to warn  about this bad usage, that it.

Reviewer: ```suggestion
Disclaimer: when we say synchronized, we don't mean two way binding. It will read initial state from URL on component mount, but after internal state will be source of truth. Which is ok most of the time, because we use `replaceState` and there is no way from user point of view to change URL without triggering page reload. The only edge case to be aware here is when you use `useURLState` with `pushState`, in this case you will need to make sure that component lives inside a valid router context.
```

> Because the URL is not really the safest way to provide state to an application, you are encouraged to provide a `decoder` function that will validate and parse the input from the URL to your state. If the input from the URL is invalid, simply return a reasonable fallback.

Reviewer: ```suggestion
Because the URL is not really the safest way to provide state to an application, you are encouraged to provide a `decoder` function that will validate and parse the input from the URL to your state. If the input from the URL is invalid, simply return a reasonable fallback or `undefined` to fall back to the default value.
```

> Because the URL is not really the safest way to provide state to an application, you need to provide an `decoder` function that validate and parse the input from the URL to your state. You need to return `undefined` if the input is not valid and `useURLState` will use default value instead.

Reviewer: > You need to return `undefined` if the input is not valid and `useURLState` will use default value instead.

Why? This is confusing.

Author: Because it's easy -> return no value if no value match.

What do you propose?

Reviewer: I propose removing the whole section about `undefined` and letting the users return fallbacks themselves.

```suggestion
Because the URL is not really the safest way to provide state to an application, you are encouraged to provide a `decoder` function that will validate and parse the input from the URL to your state. If the input from the URL is invalid, simply return a reasonable fallback.
```

It's quite convenient to be able to copy and paste a link to somebody to show them a particular page of your application.

It's even more convenient if this link has the state of the current page! This is what this hook is made for.

> Simply use `useURLState` instead of the classic `useState` and your props are automatically synchronized to the URL.

Reviewer 2: ```suggestion
Disclaimer: when we say synchronized, we don't mean two way binding. It will read initial state from URL on component mount, but after internal state will be source of truth. Which is ok most of the time, because we use `replaceState` and there is no way from user point of view to change URL without triggering page reload. The only edge case to be aware here is when you use `useURLState` with `pushState`, in this case you will need to make sure that component will re-mount on push state, for example, if you use `react-router` use component with `useURLState` inside router. 
```

@TejasQ please check wording ^

Reviewer: ```
Disclaimer: when we say synchronized, we don't mean two way binding. It will read initial state from URL on component mount, but from that point on, the component's **internal state** will be the source of truth. 

This is ok most of the time, because we use `replaceState` and there is no way from a user's point of view to change the URL without triggering page reload. 

The only edge case to be aware of here is when you use `useURLState` with `pushState`. In this case, you will need to make sure that your component will re-mount on push state, for example, if you use `react-router` use component with `useURLState` inside router.
```

> The only edge case to be aware of here is when you use `useURLState` with `pushState`. In this case, you will need to make sure that your component will re-mount on push state, for example, if you use `react-router` use component with `useURLState` inside router.

This makes no sense.

> Because the URL is not really the safest way to provide state to an application, you need to provide an `decoder` function that validate and parse the input from the URL to your state. You need to return `undefined` if the input is not valid.

Reviewer 2: You need to return `undefined` if the input is not valid and `useURLState` will use default value instead.

## Usage

```jsx

> const countDecoder = i => {

Reviewer 2: `const decode = i => Number.isNaN(+i) ? undefined : +i;`?
